[UI-side compositing] Add UI process SPIs to add page banners

shallawa opened the pull request:
#### 16b186659066d6935156abc2fe6f2fa979fa6207
<pre>
[UI-side compositing] Add UI process SPIs to add page banners
<a href="https://bugs.webkit.org/show_bug.cgi?id=253554">https://bugs.webkit.org/show_bug.cgi?id=253554</a>
rdar://101040993

Reviewed by Simon Fraser.

The client of WKWebView will provide CALayers for the page banners. Their heights
will adjust the page scroll rectangle.

Two members will be added to WebPageCreationParameters to communicate the header
and footer heights with core Page if they are are set before creating the WebPage
by the WebProcess.

RemoteLayerTreeHost::updateLayerTree() will add the banners as sub-layers to the
scrolledContentsLayer. Because it is not defined in the UIProcess, its layerID
will be added to RemoteLayerTreeTransaction.

RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree() will ensure the position
and the width of the banner layers are adjusted if the totalContentHeight or the
contentWidth change.

Mini browser will be used to test this feature manually. It will create two dummy
CALayers and call the WKWebView SPI when the setting: &quot;Reserve Space For Banners&quot;
is checked. It will call the SPI with nil pointer if this setting is unchecked.

* LayoutTests/tiled-drawing/scrolling/non-fast-region/top-content-inset-header-expected.txt:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::graphicsLayerForScrolledContents):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setHeaderHeight):
(WebCore::Page::setFooterHeight):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameTopContentInset const):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/platform/ScrollView.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::scrolledContentsLayerID const):
(WebKit::RemoteLayerTreeTransaction::setScrolledContentsLayerID):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::encode const):
(WebKit::RemoteLayerTreeTransaction::decode):
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setHeaderBannerLayer:]):
(-[WKWebView _headerBannerLayer]):
(-[WKWebView _setFooterBannerLayer:]):
(-[WKWebView _footerBannerLayer]):
* Source/WebKit/UIProcess/API/mac/WKWebViewPrivateForTestingMac.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewTestingMac.mm:
(-[WKWebView _setHeaderBannerHeight:]): Deleted.
(-[WKWebView _setFooterBannerHeight:]): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateBannerLayers):
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::topContentInset const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::layoutBannerLayers):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::setHeaderBannerHeight):
(WebKit::WebPageProxy::setFooterBannerHeight):
(WebKit::WebPageProxy::setHeaderBannerHeightForTesting): Deleted.
(WebKit::WebPageProxy::setFooterBannerHeightForTesting): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::headerBannerLayer const):
(WebKit::PageClientImpl::footerBannerLayer const):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::headerBannerLayer const):
(WebKit::WebPageProxy::footerBannerLayer const):
(WebKit::WebPageProxy::headerBannerHeight const):
(WebKit::WebPageProxy::footerBannerHeight const):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
(WebKit::WebViewImpl::headerBannerLayer const):
(WebKit::WebViewImpl::footerBannerLayer const):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setHeaderBannerLayer):
(WebKit::WebViewImpl::setFooterBannerLayer):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::setHeaderBannerHeight):
(WebKit::WebPage::setFooterBannerHeight):
(WebKit::WebPage::setHeaderBannerHeightForTesting): Deleted.
(WebKit::WebPage::setFooterBannerHeightForTesting): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree):
* Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):

Canonical link: <a href="https://commits.webkit.org/261597@main">https://commits.webkit.org/261597@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/195dd9d121a778c7bdce273cd23a999e04adb0b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112145 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21277 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3943 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120799 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4372 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117906 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99986 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105204 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98766 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/528 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45814 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13688 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/563 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14367 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19735 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52562 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8087 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16158 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2787,6 +2785,12 @@ void Page::setHeaderHeight(int headerHeight)
+
+    m_headerHeight = headerHeight;
+
+    frameView->setFrameRect(frameRect);
```
This isn't right. The size of the FrameView is the size of its frame, not the size of the contents, and banners don't affect the FrameView's size. I think setting this fixes your scrollbar bug as a side effect.

shallawa replied:
Yes you are right. We do not need to call `FrameView::setFrameRect()`. All we need is its side effect. In fact we need to call only `FrameView::availableContentSizeChanged()`.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -128,15 +128,45 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
 
+void RemoteLayerTreeDrawingAreaProxyMac::layoutBannerLayers()
+{
+    float totalContentHeight = size().height();
```
This isn't the right way to get the total height. `size()` on the DrawingAreaProxy is the size of the WKWebView, not the size of the scrolled contents. This bug should be visible if you add banners while on a long scrolling page.

The right height is probably `RemoteLayerTreeTransaction::contentsSize()` but you should figure out if this already factors in banners.

shallawa replied:
Yes this is right. `transaction.contentsSize().height()` is used instead. I tested the log scrolling page case and it worked as expected.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
```diff
@@ -97,6 +97,35 @@
+    
+    if (auto* headerBannerLayer = m_drawingArea->page().headerBannerLayer()) {
+        if ([headerBannerLayer superlayer] != scrolledContentsLayer) {
+            m_drawingArea->page().setHeaderBannerHeight(headerBannerLayer.frame.size.height);
```
Why do we need to call `setHeaderBannerHeight` here, since it's already called when setting the layers?

shallawa replied:
I added it because the client (Mini browser in my case) calls the SPIs `[WKWebView _setHeaderBannerLayer] ` before the WebPage is created by the WebProcess. So when `WebViewImpl::setHeaderBannerLayer()` calls `WebPageProxy::setHeaderBannerHeight()` the message `SetHeaderBannerHeight` is not sent to the WebProcess.

I deleted this call. Two members will be added to `WebPageCreationParameters` to communicate the header and footer heights with core `Page`.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2787,6 +2787,7 @@ void Page::setHeaderHeight(int headerHeight)
     if (!renderView)
         return;
 
+    frameView->availableContentSizeChanged(ScrollableArea::AvailableSizeChangeReason::AreaSizeChanged);
```
This isn't strictly correct; "available content size changed" means "the size of the view within which layout occurs changed" i.e. the FrameView resized, and that's not what's happening here; FrameView size does not change, but the size of the scrolled content does.

The correct fix would be to hook up a "header or footer height changed" code path and this would do work similar to what `FrameView::setContentsSize()` does. It might just enough to call `updateScrollbars(scrollPosition())`. We don't need to trigger layout, because the web contents don't change.

shallawa replied:
Fixed. Calling updateScrollbars(scrollPosition()) was enough.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -128,15 +128,43 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
+    }
+
+    if (auto* footerBannerLayer = m_webPageProxy.footerBannerLayer()) {
+        totalContentHeight += footerBannerLayer.frame.size.height;
```
Isn't this moving it too far down by the footer height?

shallawa replied:
Yes this is right. I had to use `FrameView::yPositionForHeaderLayer() `and `FrameView::yPositionForFooterLayer()` to calculate the right y position when scrolling the window.

smfr commented on Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
```diff
@@ -532,8 +532,27 @@ - (void)didChangeSettings
+        CALayer *headerBannerLayer = [[CALayer alloc] init];
+        [headerBannerLayer setBounds:CGRectMake(0, 0, 0, headerBannerHeight)];
+        [headerBannerLayer setAnchorPoint:CGPointZero];
+        [headerBannerLayer setBackgroundColor:[NSColor colorWithSRGBRed:0xac / 255. green:0xdd / 255. blue:0xde / 255. alpha:1].CGColor];
```
The mix of hex and decimal is confusing.

shallawa replied:
Fixed

shallawa committed:
[UI-side compositing] Add UI process SPIs to add page banners
https://bugs.webkit.org/show_bug.cgi?id=253554
rdar://101040993

Reviewed by Simon Fraser.

The client of WKWebView will provide CALayers for the page banners. Their heights
will adjust the page scroll rectangle.

Two members will be added to WebPageCreationParameters to communicate the header
and footer heights with core Page if they are are set before creating the WebPage
by the WebProcess.

RemoteLayerTreeHost::updateLayerTree() will add the banners as sub-layers to the
scrolledContentsLayer. Because it is not defined in the UIProcess, its layerID
will be added to RemoteLayerTreeTransaction.

RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree() will ensure the position
and the width of the banner layers are adjusted if the totalContentHeight or the
contentWidth change.

Mini browser will be used to test this feature manually. It will create two dummy
CALayers and call the WKWebView SPI when the setting: "Reserve Space For Banners"
is checked. It will call the SPI with nil pointer if this setting is unchecked.

* LayoutTests/tiled-drawing/scrolling/non-fast-region/top-content-inset-header-expected.txt:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::graphicsLayerForScrolledContents):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setHeaderHeight):
(WebCore::Page::setFooterHeight):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameTopContentInset const):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/platform/ScrollView.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::scrolledContentsLayerID const):
(WebKit::RemoteLayerTreeTransaction::setScrolledContentsLayerID):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::encode const):
(WebKit::RemoteLayerTreeTransaction::decode):
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setHeaderBannerLayer:]):
(-[WKWebView _headerBannerLayer]):
(-[WKWebView _setFooterBannerLayer:]):
(-[WKWebView _footerBannerLayer]):
* Source/WebKit/UIProcess/API/mac/WKWebViewPrivateForTestingMac.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewTestingMac.mm:
(-[WKWebView _setHeaderBannerHeight:]): Deleted.
(-[WKWebView _setFooterBannerHeight:]): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateBannerLayers):
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::topContentInset const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::layoutBannerLayers):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::setHeaderBannerHeight):
(WebKit::WebPageProxy::setFooterBannerHeight):
(WebKit::WebPageProxy::setHeaderBannerHeightForTesting): Deleted.
(WebKit::WebPageProxy::setFooterBannerHeightForTesting): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::headerBannerLayer const):
(WebKit::PageClientImpl::footerBannerLayer const):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::headerBannerLayer const):
(WebKit::WebPageProxy::footerBannerLayer const):
(WebKit::WebPageProxy::headerBannerHeight const):
(WebKit::WebPageProxy::footerBannerHeight const):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
(WebKit::WebViewImpl::headerBannerLayer const):
(WebKit::WebViewImpl::footerBannerLayer const):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setHeaderBannerLayer):
(WebKit::WebViewImpl::setFooterBannerLayer):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::setHeaderBannerHeight):
(WebKit::WebPage::setFooterBannerHeight):
(WebKit::WebPage::setHeaderBannerHeightForTesting): Deleted.
(WebKit::WebPage::setFooterBannerHeightForTesting): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree):
* Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):

Canonical link: https://commits.webkit.org/261597@main
